flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SlotSharingExecutionSlotAllocator.java

```diff
@@ -160,38 +162,58 @@ private void cancelLogicalSlotRequest(ExecutionVertexID executionVertexId, Throw
-
-			for (ExecutionVertexID executionId : executionIds) {
-				CompletableFuture<LogicalSlot> logicalSlotFuture = sharedSlot.allocateLogicalSlot(executionId);
-				SlotExecutionVertexAssignment assignment = new SlotExecutionVertexAssignment(executionId, logicalSlotFuture);
-				assignments.put(executionId, assignment);
-			}
+			allocateLogicalSlotsFromSharedSlot(assignments, executionIds, sharedSlot);
 		}
 
 		return assignments;
 	}
 
 	private SharedSlot getOrAllocateSharedSlot(
-			ExecutionSlotSharingGroup executionSlotSharingGroup,
+			ExecutionSlotSharingGroup slotSharingGroup,
 			SharedSlotProfileRetriever sharedSlotProfileRetriever) {
-		return sharedSlots
-			.computeIfAbsent(executionSlotSharingGroup, group -> {
-				SlotRequestId physicalSlotRequestId = new SlotRequestId();
-				ResourceProfile physicalSlotResourceProfile = getPhysicalSlotResourceProfile(group);
-				SlotProfile slotProfile = sharedSlotProfileRetriever.getSlotProfile(group, physicalSlotResourceProfile);
-				PhysicalSlotRequest physicalSlotRequest =
-					new PhysicalSlotRequest(physicalSlotRequestId, slotProfile, slotWillBeOccupiedIndefinitely);
-				CompletableFuture<PhysicalSlot> physicalSlotFuture = slotProvider
+		SharedSlot sharedSlot = sharedSlots.get(slotSharingGroup);
+		if (sharedSlot == null) {
+			SlotRequestId physicalSlotRequestId = new SlotRequestId();
+			ResourceProfile physicalSlotResourceProfile =
+					getPhysicalSlotResourceProfile(slotSharingGroup);
+			SlotProfile slotProfile = sharedSlotProfileRetriever
+					.getSlotProfile(slotSharingGroup, physicalSlotResourceProfile);
+			PhysicalSlotRequest physicalSlotRequest = new PhysicalSlotRequest(
+					physicalSlotRequestId,
+					slotProfile,
+					slotWillBeOccupiedIndefinitely);
+			CompletableFuture<PhysicalSlot> physicalSlotFuture = slotProvider
 					.allocatePhysicalSlot(physicalSlotRequest)
 					.thenApply(PhysicalSlotRequest.Result::getPhysicalSlot);
-				return new SharedSlot(
+			sharedSlot = new SharedSlot(
 					physicalSlotRequestId,
 					physicalSlotResourceProfile,
-					group,
+					slotSharingGroup,
 					physicalSlotFuture,
 					slotWillBeOccupiedIndefinitely,
 					this::releaseSharedSlot);
-			});
+			if (!physicalSlotFuture.isCompletedExceptionally()) {
+				sharedSlots.put(slotSharingGroup, sharedSlot);
+			}
+		}
+		return sharedSlot;
+	}
+
+	private static void allocateLogicalSlotsFromSharedSlot(
+			Map<ExecutionVertexID, SlotExecutionVertexAssignment> assignments,
+			Iterable<ExecutionVertexID> executionIds,
+			SharedSlot sharedSlot) {
+		for (ExecutionVertexID executionId : executionIds) {
+			boolean physicalSlotHasAlreadyFailed = sharedSlot
+					.getSlotContextFuture()
+					.isCompletedExceptionally();
+			CompletableFuture<LogicalSlot> logicalSlotFuture = physicalSlotHasAlreadyFailed ?
+					sharedSlot.getSlotContextFuture().thenApply(stub -> null) :
+					sharedSlot.allocateLogicalSlot(executionId);
```

Comment:
can't `SharedSlot.allocateLogicalSlot` return a failed future if the internal future has already failed? That way, we wouldn't need the special casing here.